core/src/main/java/com/linecorp/armeria/common/Flags.java

```diff
@@ -1422,87 +1304,70 @@ private static ExceptionVerbosity exceptionLoggingMode(String name, String defau
-        final String value = System.getProperty(fullName);
-        if (value == null) {
-            logger.info("{}: {} (default)", fullName, defaultValue);
-            return defaultValue;
+        String value = System.getProperty(fullName);
+        if (value != null) {
+            value = Ascii.toLowerCase(value);
         }
 
-        if (validator.test(value)) {
-            logger.info("{}: {}", fullName, value);
-            return value;
+        if (value != null) {
+            if (validator.test(value)) {
+                logger.info("{}: {} (jvm option)", fullName, value);
+                return value;
+            }
+            logger.warn("{}: {} (jvm option) fail validation", fullName, value);
         }
-
-        logger.info("{}: {} (default instead of: {})", fullName, defaultValue, value);
+        logger.info("{}: {} (default)", fullName, defaultValue);
         return defaultValue;
     }
 
-    private static String getNormalized(String name, String defaultValue, Predicate<String> validator) {
-        final String fullName = PREFIX + name;
-        final String value = getLowerCased(fullName);
-        if (value == null) {
-            logger.info("{}: {} (default)", fullName, defaultValue);
-            return defaultValue;
-        }
+    private static <T> T getValue(Function<FlagsProvider, @Nullable T> method, String flagName) {
+        return getValue(method, flagName, unused -> true);
+    }
 
-        if (validator.test(value)) {
-            logger.info("{}: {}", fullName, value);
-            return value;
-        }
+    private static <T> T getValue(Function<FlagsProvider, @Nullable T> method,
+                                  String flagName, Predicate<T> validator) {
+        for (FlagsProvider provider : FLAGS_PROVIDERS) {
+            T value = null;
+            try {
+                value = method.apply(provider);
+            } catch (Exception ex) {
+                logger.warn("{}: ({}) fail to get value, {}", flagName, provider.getClass().getSimpleName(),
+                            ex.getMessage());
+            }
 
-        logger.info("{}: {} (default instead of: {})", fullName, defaultValue, value);
-        return defaultValue;
-    }
+            if (value == null) {
+                continue;
+            }
 
-    @Nullable
-    private static String getLowerCased(String fullName) {
-        String value = System.getProperty(fullName);
-        if (value != null) {
-            value = Ascii.toLowerCase(value);
+            if (!validator.test(value)) {
+                logger.warn("{}: {} ({}) fail validation",
+                            flagName, value, provider.getClass().getSimpleName());
+                continue;
+            }
+
+            logger.info("{}: {} ({})", flagName, value, provider.getClass().getSimpleName());
+            return value;
         }
-        return value;
+        final T defaultValue = method.apply(DefaultFlagsProvider.INSTANCE);
```

Comment:
Should we append `DefaultFlagsProvider` to `FLAGS_PROVIDERS` as the last provider? We can use the last value if there is no valid value.

Reply:
I could do that. Thanks.